### PR TITLE
Fix Stack Overflow Error when reading from partition table

### PR DESCRIPTION
**Pull request by marsishandsome**
### What problem does this PR solve? 
close https://github.com/pingcap/tispark/issues/1181
ref https://asktug.com/t/tispark-count/1604

Stack Overflow Error when reading from partition table

org.apache.spark.rdd.RDD.preferredLocations(RDD.scala:274) at org.apache.spark.rdd.UnionPartition.preferredLocations(UnionRDD.scala:49) at org.apache.spark.rdd.UnionRDD.getPreferredLocations(UnionRDD.scala:109) at org.apache.spark.rdd.RDD$$anonfun$preferredLocations$2.apply(RDD.scala:275) at org.apache.spark.rdd.RDD$$anonfun$preferredLocations$2.apply(RDD.scala:275) at scala.Option.getOrElse(Option.scala:121) at org.apache.spark.rdd.RDD.preferredLocations(RDD.scala:274) at org.apache.spark.rdd.UnionPartition.preferredLocations(UnionRDD.scala:49) at org.apache.spark.rdd.UnionRDD.getPreferredLocations(UnionRDD.scala:109) at org.apache.spark.rdd.RDD$$anonfun$preferredLocations$2.apply(RDD.scala:275) at org.apache.spark.rdd.RDD$$anonfun$preferredLocations$2.apply(RDD.scala:275) at scala.Option.getOrElse(Option.scala:121) at org.apache.spark.rdd.RDD.preferredLocations(RDD.scala:274) at org.apache.spark.rdd.UnionPartition.preferredLocations(UnionRDD.scala:49) at org.apache.spark.rdd.UnionRDD.getPreferredLocations(UnionRDD.scala:109) at org.apache.spark.rdd.RDD$$anonfun$preferredLocations$2.apply(RDD.scala:275) at org.apache.spark.rdd.RDD$$anonfun$preferredLocations$2.apply(RDD.scala:275) at scala.Option.getOrElse(Option.scala:121) at org.apache.spark.rdd.RDD.preferredLocations(RDD.scala:274) at org.apache.spark.rdd.UnionPartition.preferredLocations(UnionRDD.scala:49) at org.apache.spark.rdd.UnionRDD.getPreferredLocations(UnionRDD.scala:109) at org.apache.spark.rdd.RDD$$anonfun$preferredLocations$2.apply(RDD.scala:275) at org.apache.spark.rdd.RDD$$anonfun$preferredLocations$2.apply(RDD.scala:275) at scala.Option.getOrElse(Option.scala:121) at org.apache.spark.rdd.RDD.preferredLocations(RDD.scala:274) at org.apache.spark.rdd.UnionPartition.preferredLocations(UnionRDD.scala:49) at org.apache.spark.rdd.UnionRDD.getPreferredLocations(UnionRDD.scala:109) at org.apache.spark.rdd.RDD$$anonfun$preferredLocations$2.apply(RDD.scala:275) at org.apache.spark.rdd.RDD$$anonfun$preferredLocations$2.apply(RDD.scala:275) at scala.Option.getOrElse(Option.scala:121) at org.apache.spark.rdd.RDD.preferredLocations(RDD.scala:274) at org.apache.spark.rdd.UnionPartition.preferredLocations(UnionRDD.scala:49) at org.apache.spark.rdd.UnionRDD.getPreferredLocations(UnionRDD.scala:109) at org.apache.spark.rdd.RDD$$anonfun$preferredLocations$2.apply(RDD.scala:275) at org.apache.spark.rdd.RDD$$anonfun$preferredLocations$2.apply(RDD.scala:275) at scala.Option.getOrElse(Option.scala:121) at org.apache.spark.rdd.RDD.preferredLocations(RDD.scala:274) at org.apache.spark.rdd.UnionPartition.preferredLocations(UnionRDD.scala:49) at org.apache.spark.rdd.UnionRDD.getPreferredLocations(UnionRDD.scala:109) at org.apache.spark.rdd.RDD$$anonfun$preferredLocations$2.apply(RDD.scala:275) at org.apache.spark.rdd.RDD$$anonfun$preferredLocations$2.apply(RDD.scala:275) at scala.Option.getOrElse(Option.scala:121) at org.apache.spark.rdd.RDD.preferredLocations(RDD.scala:274) at org.apache.spark.rdd.UnionPartition.preferredLocations(UnionRDD.scala:49) at org.apache.spark.rdd.UnionRDD.getPreferredLocations(UnionRDD.scala:109) at org.apache.spark.rdd.RDD$$anonfun$preferredLocations$2.apply(RDD.scala:275) at org.apache.spark.rdd.RDD$$anonfun$preferredLocations$2.apply(RDD.scala:275) at scala.Option.getOrElse(Option.scala:121) at org.apache.spark.rdd.RDD.preferredLocations(RDD.scala:274) at org.apache.spark.rdd.UnionPartition.preferredLocations(UnionRDD.scala:49) at org.apache.spark.rdd.UnionRDD.getPreferredLocations(UnionRDD.scala:109) at org.apache.spark.rdd.RDD$$anonfun$preferredLocations$2.apply(RDD.scala:275) at org.apache.spark.rdd.RDD$$anonfun$preferredLocations$2.apply(RDD.scala:275) at scala.Option.getOrElse(Option.scala:121) at org.apache.spark.rdd.RDD.preferredLocations(RDD.scala:274) at org.apache.spark.rdd.UnionPartition.preferredLocations(UnionRDD.scala:49) at org.apache.spark.rdd.UnionRDD.getPreferredLocations(UnionRDD.scala:109) at org.apache.spark.rdd.RDD$$anonfun$preferredLocations$2.apply(RDD.scala:275) at org.apache.spark.rdd.RDD$$anonfun$preferredLocations$2.apply(RDD.scala:275) at scala.Option.getOrElse(Option.scala:121)

### What is changed and how it works?
use `SparkContext.union` instead of `RDD.union`

### Check List 

Tests 

 - Unit test

Related changes

 - Need to cherry-pick to the release branch
 - Need to be included in the release note

### CODE REVIEW

**Review by birdstorm — APPROVED** [review]
LGTM

File: core/src/test/scala/org/apache/spark/sql/PartitionTableSuite.scala [hunk 1]
```diff
@@ -21,6 +21,50 @@ import org.apache.spark.sql.execution.{CoprocessorRDD, RegionTaskExec}
...
+    }
+
+    {
+      // no index scan
+      tidbStmt.execute("DROP TABLE IF EXISTS `pt`")
+      tidbStmt.execute(s"""
+                          |CREATE TABLE `pt` (
+                          |  `id` int(11) DEFAULT NULL,
+                          |  `name` varchar(50) DEFAULT NULL,
+                          |  `purchased` date DEFAULT NULL
```
Comment (zhexuany) [inline core/src/test/scala/org/apache/spark/sql/PartitionTableSuite.scala]: this test case only covers index scan case. We can add another test case to cover table scan case.

**Review by ilovesoup — APPROVED** [review]
LGTM

File: core/src/test/scala/org/apache/spark/sql/PartitionTableSuite.scala (hunk 1 above)
Comment (marsishandsome) [reply core/src/test/scala/org/apache/spark/sql/PartitionTableSuite.scala]: see line 29-45

**Pull request merged**